modules/viewability.js

```diff
@@ -0,0 +1,163 @@
+import { logWarn, logInfo, isStr, isFn, triggerPixel, insertHtmlIntoIframe } from '../src/utils.js';
+import { getGlobal } from '../src/prebidGlobal.js';
+import find from 'core-js-pure/features/array/find.js';
+
+export function init() {
+  (getGlobal()).viewability = {
+    startMeasurement: startMeasurement,
+    stopMeasurement: stopMeasurement,
+  };
+
+  listenMessagesFromCreative();
+}
+
+const observers = {};
+
+function isValid(vid, element, tracker, criteria) {
+  if (!element) {
+    logWarn('provide an html element to track');
+    return false;
+  }
+
+  let validTracker = tracker &&
+    ((tracker.method === 'img' && isStr(tracker.value)) ||
+    (tracker.method === 'js' && isStr(tracker.value)) ||
+    (tracker.method === 'callback' && isFn(tracker.value)));
+
+  if (!validTracker) {
+    logWarn('invalid tracker', tracker);
+    return false;
+  }
+
+  if (!criteria || !criteria.inViewThreshold || !criteria.timeInView) {
+    logWarn('missing criteria', criteria);
+    return false;
+  }
+
+  if (!vid || observers[vid]) {
+    logWarn('provide an unregistered vid', vid);
+    return false;
+  }
+
+  return true;
+}
+
+/**
+ * Start measuring viewability of an element
+ * @typedef {{ method: string='img','js','callback', value: string|function }} ViewabilityTracker { method: 'img', value: 'http://my.tracker/123' }
+ * @typedef {{ inViewThreshold: number, timeInView: number }} ViewabilityCriteria { inViewThreshold: 0.5, timeInView: 1000 }
+ * @param {string} vid unique viewability identifier
+ * @param {HTMLElement} element
+ * @param {ViewabilityTracker} tracker
+ * @param {ViewabilityCriteria} criteria
+ */
+export function startMeasurement(vid, element, tracker, criteria) {
+  if (!isValid(vid, element, tracker, criteria)) {
+    return;
+  }
+
+  let options = {
+    root: null,
+    rootMargin: '0px',
+    threshold: criteria.inViewThreshold,
+  };
+
+  let observer;
+  let viewable = false;
+  let stateChange = (entries) => {
+    viewable = entries[0].isIntersecting;
+
+    if (viewable) {
+      observers[vid].timeoutId = window.setTimeout(() => {
+        // stop observing
+        observer.unobserve(element);
+        observers[vid].done = true;
+
+        switch (tracker.method) {
+          case 'img':
+            triggerPixel(tracker.value, () => {
+              logInfo('viewability pixel fired', tracker.value);
+            });
+            break;
+          case 'js':
+            insertHtmlIntoIframe(`<script src="${tracker.value}"></script>`);
+            break;
+          case 'callback':
+            tracker.value(element);
+            break;
+        }
+      }, criteria.timeInView);
+    } else if (observers[vid].timeoutId) {
+      window.clearTimeout(observers[vid].timeoutId);
+    }
+  };
+
+  observer = new IntersectionObserver(stateChange, options);
```

Comment:
I was going through the docs for IntersectionObserver API on MDN, on the [Browser compatibility section](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API#browser_compatibility), I found that Safari and few of the other Mobile browsers don't support `options.root`. I'm not sure if it'll throw and error and crash on those browsers, do we need a fallback or some error handling?

Reply:
I assume your referring to this `options.root parameter can be a Document`? If that's the case, we're setting options.root to null in this PR, not to any document?